### DISCUSSION TITLE
More in-depth examples on message passing?

### DISCUSSION THREAD

**Discussion by igorfil** [Textualize/textual#846] [Q&A]
Hi,

I really love the Textual! One thing I would like to understand better is message passing. 

There are not many examples really and I did not find documentation. From the discussions and the Calculator example it is more or less clear how to pass messages between widgets. What I would like to know is how to pass messages into the Textual App from the outside? I.e. when Textual is only a frontend for a bigger application, possibly running multiple threads. 

For example, I have a thread that processes some data, and when it's done, I would like it to be displayed. I would like to understand how to pass a message with data to be displayed into the Textual App and then channel it to a proper Widget that will render it.

Is there any examples that would show proper ways of doing that?

**Comment by zed** [comment]
As I see it, builtin messages are used by widgets to bubble up events such as `Button.Pressed` https://textual.textualize.io/guide/events/
reactive/var can be used by a parent widget for the opposite top-down direction. https://textual.textualize.io/guide/reactivity/
There is also experimental (for internal use) `textual.signal`: having a signal, widgets can subscribe; signal owner publishes e.g., `app.theme_changed_signal`. 

More generally, [event-related Python packages](https://stackoverflow.com/a/16192256/4279) follow Observer/Publish-subscribe patterns (there is a dispatch/message bus object in the latter case). For example, your UI code responsible for displaying `model.some_attr` can connect to changes in your model object: `model.events.some_attr.connect(callback)` and the code responsible for updating the value: `model.some_attr = new_value`--assuming something like https://pypi.org/project/psygnal/ (similar to [Qt's Signals & Slots API](https://doc.qt.io/qt-5/signalsandslots.html)). To handle multiple threads, there may be additional steps such as [`.connect(thread="current")` & `.emit_queued()`](https://psygnal.readthedocs.io/en/stable/usage/?h=threads#connecting-across-threads).
